src/log_surgeon/finite_automata/RegexNFA.hpp

```diff
@@ -179,44 +183,46 @@ class RegexNFA {
 public:
     using StateVec = std::vector<NFAStateType*>;
 
-    explicit RegexNFA(std::vector<LexicalRule<NFAStateType>> const& m_rules);
+    explicit RegexNFA(std::vector<LexicalRule<NFAStateType>> rules);
 
     /**
-     * Create a unique_ptr for an NFA state with no tagged transitions and add it to m_states.
+     * Creates a unique_ptr for an NFA state with no tagged transitions and adds it to `m_states`.
      * @return NFAStateType*
      */
-    auto new_state() -> NFAStateType*;
+    [[nodiscard]] auto new_state() -> NFAStateType*;
 
     /**
-     * Create a unique_ptr for an NFA state with a positive tagged transition and add it to
-     * m_states.
+     * Creates a unique_ptr for an NFA state with a positive tagged transition and adds it to
+     * `m_states`.
      * @param tag
      * @param dest_state
      * @return NFAStateType*
      */
-    auto new_state_with_positive_tagged_transition(uint32_t tag, NFAStateType const* dest_state)
-            -> NFAStateType*;
+    [[nodiscard]] auto new_state_with_positive_tagged_transition(
+            uint32_t tag,
+            NFAStateType const* dest_state
+    ) -> NFAStateType*;
 
     /**
-     * Create a unique_ptr for an NFA state with negative tagged transitions and add it to m_states.
-     * @param tags
+     * Creates a unique_ptr for an NFA state with negative tagged transitions and adds it to
+     * `m_states`.
      * @param dest_state
      * @return NFAStateType*
      */
-    auto new_state_with_negative_tagged_transitions(
+    [[nodiscard]] auto new_state_with_negative_tagged_transitions(
             std::set<uint32_t> tags,
             NFAStateType const* dest_state
     ) -> NFAStateType*;
 
     /**
-     * Traverse the NFA using a BFS and keep track of the order states are visited in.
-     * @return A vector that performs a BFS of the NFA.
+     * Traverse the NFA using a breadth-first search (BFS) and keep track of the order states are
+     * visited in.
+     * @return A vector representing the traversal order of the NFA states using BFS.
      */
-    [[nodiscard]] auto get_traversal_order() const -> std::vector<RegexNFAByteState const*>;
+    [[nodiscard]] auto get_bfs_traversal_order() const -> std::vector<RegexNFAByteState const*>;
 
     /**
-     * Serialize the NFA into a string.
-     * @return A string representation of the NFA.
+     * @return A string representation of the NFA. This function should always succeeed.
```

Comment:
```suggestion
     * @return A string representation of the NFA.
```
We don't have to explain this right? If the return value isn't an optional, then it is assumed that it won't fail; if it would throw (which will never happen in log-surgeon), we should add a docstring for which exception it would throw.

Reply:
From a doc string perspective, it is showing user who calls it how this function does. And if the return value doesn't have any error propagated, or there's no `throw` section, then it should be assumed "always success".
In terms of differentiating with other methods, I think u're right that the internal comments are sufficient.